### PR TITLE
Convert new SigLIP checkpoints

### PR DESCRIPTION
**Pull request by merveenoyan**
@NielsRogge as requested, I have modified the conversion script to convert new SigLIP ckpts.

I have modified model max length and pos embed shape because normally it was 64 and for the so-400m-patch14-224 it is now 16. 
My problem is that the expected output and the converted model outputs aren't close enough, see below. Did you come across the same issue?

```
outputs.logits_per_image[:3, :3]: tensor([[-1.0865,  1.1704],
        [-0.7178,  1.4354]])
expected_slice: tensor([[-1.0837,  1.1743],
        [-0.7196,  1.4371]])
```

Skipping the other model for now, the vocab dim of text embeddings is a bit of an issue, it is said to be 250k but is 256k, waiting for response from google folks (I think they used a different tokenizer, the demo notebook of big-vision also doesn't work for this reason)

Update: I confirmed with folks that there seems to be a mistake with tokenizer.

### CODE REVIEW

File: src/transformers/models/siglip/convert_siglip_to_hf.py [hunk 1]
```diff
@@ -49,6 +49,8 @@
     "siglip-base-patch16-256-i18n": "/Users/nielsrogge/Documents/SigLIP/webli_i18n_b16_256_66117334.npz",
     # so400m checkpoints
     "siglip-so400m-patch14-384": "/Users/nielsrogge/Documents/SigLIP/webli_en_so400m_384_58765454.npz",
+    "siglip-so400m-patch14-224": "/Users/mervenoyan/Desktop/siglips/webli_en_so400m_224_57633886.npz",
+#    "siglip-so400m-patch16-256": "/Users/mervenoyan/Desktop/siglips/webli_i18n_so400m_16_256_78061115.npz",
```
Comment (lucasb-eyer) [inline src/transformers/models/siglip/convert_siglip_to_hf.py:53]: We should have -i18n in the model name, my fault for forgetting this in our colab, updating soon.

Comment (lucasb-eyer) [reply src/transformers/models/siglip/convert_siglip_to_hf.py:53]: Updated our colab.

Comment (NielsRogge) [reply src/transformers/models/siglip/convert_siglip_to_hf.py:53]: To be uncommented?

File: src/transformers/models/siglip/convert_siglip_to_hf.py [hunk 2]
```diff
@@ -97,6 +101,9 @@ def get_siglip_config(model_name):
         config.vision_config.num_attention_heads = 16
     else:
         raise ValueError("Model not supported")
+    
+    if model_name == "siglip-so400m-patch14-224":
```
Comment (NielsRogge) [inline src/transformers/models/siglip/convert_siglip_to_hf.py:105]: This can go above the "Model not supported"

File: src/transformers/models/siglip/convert_siglip_to_hf.py [hunk 3]
```diff
@@ -266,9 +273,9 @@ def convert_siglip_checkpoint(model_name, pytorch_dump_folder_path, verify_logit
 
     # get vocab file
     if "i18n" in model_name:
-        vocab_file = "/Users/nielsrogge/Documents/SigLIP/multilingual_vocab/sentencepiece.model"
+        vocab_file = "/Users/mervenoyan/Desktop/siglips/mc4/sentencepiece.model"
```
Comment (NielsRogge) [inline src/transformers/models/siglip/convert_siglip_to_hf.py:276]: We could host these on the hub rather than having local paths so that everyone can run it

File: src/transformers/models/siglip/convert_siglip_to_hf.py [hunk 4]
```diff
@@ -285,15 +292,18 @@ def convert_siglip_checkpoint(model_name, pytorch_dump_folder_path, verify_logit
...
     # create processor
     # important: make tokenizer not return attention_mask since original one doesn't require it
     image_size = config.vision_config.image_size
     size = {"height": image_size, "width": image_size}
     image_processor = SiglipImageProcessor(size=size)
     tokenizer = SiglipTokenizer(vocab_file=vocab_file, model_input_names=["input_ids"])
-    processor = SiglipProcessor(image_processor=image_processor, tokenizer=tokenizer)
 
+    if model_name == "siglip-so400m-patch14-224":
+        tokenizer.model_max_length = 16
```
Comment (NielsRogge) [inline src/transformers/models/siglip/convert_siglip_to_hf.py:303]: I don't think we set the `model_max_length` attribute for the existing checkpoints, should we do that? Would it break anything?

**Comment by merveenoyan** [comment]
Sorry for the delay, I was busy shipping Llama 3.2 release and then ECCV and then I took time off. I just got back to this PR.

File: src/transformers/models/siglip/convert_siglip_to_hf.py (hunk 4 above)
Comment (merveenoyan) [reply src/transformers/models/siglip/convert_siglip_to_hf.py:303]: for this one it is different, that's why I changed it

File: src/transformers/models/siglip/convert_siglip_to_hf.py (hunk 3 above)
Comment (merveenoyan) [reply src/transformers/models/siglip/convert_siglip_to_hf.py:276]: Sure I'll upload them all.

Comment (merveenoyan) [reply src/transformers/models/siglip/convert_siglip_to_hf.py:276]: Should we have them in same repo as the HF checkpoints? @NielsRogge

File: src/transformers/models/siglip/convert_siglip_to_hf.py (hunk 1 above)
Comment (merveenoyan) [reply src/transformers/models/siglip/convert_siglip_to_hf.py:53]: Yep I'm porting it

**Comment by merveenoyan** [comment]
This is going to have more changes in the script than I expected bcs previously we had e.g. `img/Transformer/encoderblock_{i}/MlpBlock_0/Dense_1/kernel` where i was a hidden layer order, whereas in here `img/Transformer/encoderblock/MlpBlock_0/Dense_1/kernel` contains same var for all 27 hidden layers as array of tensors, so I will have to change more things

File: src/transformers/models/siglip/modeling_siglip.py [hunk 5]
```diff
@@ -926,7 +926,7 @@ def __init__(self, config: SiglipTextConfig):
         self.embeddings = SiglipTextEmbeddings(config)
         self.encoder = SiglipEncoder(config)
         self.final_layer_norm = nn.LayerNorm(embed_dim, eps=config.layer_norm_eps)
-
+        #if not (config.vision_config.image_size==256 and config.text_config.vocab_size==250000 and config.vision_config.patch_size==16):
```
Comment (merveenoyan) [inline src/transformers/models/siglip/modeling_siglip.py:929]: @NielsRogge we cannot check the vision config here because `TextTransformer` is constructed from text config. We need to add an attr to the config otherwise there's no other way of checking this.

These are the specs of this new model BTW that's why I'm specifically checking them, but I find it a bit hacky and it won't be possible unless I change it to take in complete config but this is against the patterns so I wouldn't like to do that.

File: src/transformers/models/siglip/convert_siglip_to_hf.py [hunk 6]
```diff
@@ -263,37 +273,54 @@ def convert_siglip_checkpoint(model_name, pytorch_dump_folder_path, verify_logit
...
     else:
-        vocab_file = "/Users/nielsrogge/Documents/SigLIP/english_vocab/sentencepiece.model"
+        vocab_file = "/Users/mervenoyan/Desktop/siglips/c4_en/sentencepiece.model"
 
     # load original state dict
     data = load(checkpoint)
     state_dict = flatten_nested_dict(data)
 
-    # remove and rename some keys
+    if siglip_sovit_i18_256:  # make state dict compatible with rest of the SiglIPs, add param/ prefix and encoderblock index
```
Comment (merveenoyan) [inline src/transformers/models/siglip/convert_siglip_to_hf.py:289]: this model has a single `encoderblock` and not `encoderblock_i` for all hidden layers in vision tower like others so I modified the state dict to be compatible with the rest of the modelling code  @NielsRogge

File: src/transformers/models/siglip/convert_siglip_to_hf.py [hunk 7]
```diff
@@ -167,8 +176,9 @@ def create_rename_keys(config):
 
     rename_keys.append(("params/txt/Encoder_0/encoder_norm/scale", "text_model.final_layer_norm.weight"))
     rename_keys.append(("params/txt/Encoder_0/encoder_norm/bias", "text_model.final_layer_norm.bias"))
-    rename_keys.append(("params/txt/head/kernel", "text_model.head.weight"))
-    rename_keys.append(("params/txt/head/bias", "text_model.head.bias"))
+    if not siglip_sovit_i18_256:
```
Comment (merveenoyan) [inline src/transformers/models/siglip/convert_siglip_to_hf.py:179]: This model doesn't have a head for text part

**Comment by merveenoyan** [comment]
The logits are validated so except for cosmetic changes we are good to merge✨ the model is here https://huggingface.co/merve/siglip-so400m-patch16-256-i18n

File: src/transformers/models/siglip/convert_siglip_to_hf.py [hunk 8]
```diff
@@ -95,6 +100,14 @@ def get_siglip_config(model_name):
...
         config.vision_config.num_hidden_layers = 27
         config.vision_config.num_attention_heads = 16
+    if "so400m-patch14-224" in model_name:
+        config.text_config.max_position_embeddings = 16
+    if (
+        config.vision_config.image_size == 256
+        and config.text_config.vocab_size == 250000
+        and config.vision_config.patch_size == 16
+    ):
+        config.text_config.no_head = True
```
Comment (NielsRogge) [inline src/transformers/models/siglip/convert_siglip_to_hf.py:110]:
```suggestion
    elif model_name == "siglip-so400m-patch16-256-i18n":
        config.text_config.no_head = True
```

File: src/transformers/models/siglip/convert_siglip_to_hf.py [hunk 9]
```diff
@@ -263,35 +278,58 @@ def convert_siglip_checkpoint(model_name, pytorch_dump_folder_path, verify_logit
 
     # get checkpoint
     checkpoint = model_name_to_checkpoint[model_name]
+    if (
+        config.vision_config.image_size == 256
+        and config.text_config.vocab_size == 250000
+        and config.vision_config.patch_size == 16
+    ):
+        siglip_sovit_i18_256 = True
```
Comment (NielsRogge) [inline src/transformers/models/siglip/convert_siglip_to_hf.py:286]: This can be removed in favor of `if model_name == "siglip-so400m-patch16-256-i18n"`

File: src/transformers/models/siglip/convert_siglip_to_hf.py (hunk 3 above)
Comment (NielsRogge) [reply src/transformers/models/siglip/convert_siglip_to_hf.py:290]: No, they can be uploaded to a separate repo

File: src/transformers/models/siglip/convert_siglip_to_hf.py (hunk 9 above)
Comment (merveenoyan) [reply src/transformers/models/siglip/convert_siglip_to_hf.py]: I removed this part and `siglip_sovit_i18_256` checks altogether since I was checking it in only one line

File: src/transformers/models/siglip/convert_siglip_to_hf.py [hunk 10]
```diff
@@ -308,6 +310,7 @@ def convert_siglip_checkpoint(model_name, pytorch_dump_folder_path, verify_logit
     read_in_q_k_v_head(state_dict, config)
     # load HuggingFace model
     model = SiglipModel(config).eval()
+    print("config.text_config", config.text_config)
```
Comment (NielsRogge) [inline src/transformers/models/siglip/convert_siglip_to_hf.py]: Can be removed

Comment (merveenoyan) [reply src/transformers/models/siglip/convert_siglip_to_hf.py]: it's outdated I think

File: src/transformers/models/siglip/modeling_siglip.py [hunk 11]
```diff
@@ -926,8 +926,8 @@ def __init__(self, config: SiglipTextConfig):
         self.embeddings = SiglipTextEmbeddings(config)
         self.encoder = SiglipEncoder(config)
         self.final_layer_norm = nn.LayerNorm(embed_dim, eps=config.layer_norm_eps)
-        #if not (config.vision_config.image_size==256 and config.text_config.vocab_size==250000 and config.vision_config.patch_size==16):
-        self.head = nn.Linear(embed_dim, embed_dim)
+        if not hasattr(self.config, "no_head"):
```
Comment (NielsRogge) [inline src/transformers/models/siglip/modeling_siglip.py:929]: This should be a boolean attribute, which is cleaner to use that `hasattr`.

I don't see updates to configuration_siglip.py, the no_head argument should be added there

Comment (merveenoyan) [reply src/transformers/models/siglip/modeling_siglip.py]: addressed

**Comment by merveenoyan** [comment]
I keep the model repo in sync with the changes here

File: src/transformers/models/siglip/convert_siglip_to_hf.py [hunk 12]
```diff
@@ -95,6 +100,10 @@ def get_siglip_config(model_name):
         config.vision_config.intermediate_size = 4304
         config.vision_config.num_hidden_layers = 27
         config.vision_config.num_attention_heads = 16
+    if "so400m-patch14-224" in model_name:
```
Comment (NielsRogge) [inline src/transformers/models/siglip/convert_siglip_to_hf.py:103]:
```suggestion
    if model_name == "siglip-so400m-patch14-224":
```

File: src/transformers/models/siglip/modeling_siglip.py [hunk 13]
```diff
@@ -978,7 +978,8 @@ def forward(
 
         # Assuming "sticky" EOS tokenization, last token is always EOS.
         pooled_output = last_hidden_state[:, -1, :]
-        pooled_output = self.head(pooled_output)
+        if self.config.has_head:
+            pooled_output = self.head(pooled_output)
```
Comment (NielsRogge) [inline src/transformers/models/siglip/modeling_siglip.py:982]:
```suggestion
        if self.config.has_head is not None:
            pooled_output = self.head(pooled_output)
```

File: src/transformers/models/siglip/modeling_siglip.py [hunk 14]
```diff
@@ -926,8 +926,8 @@ def __init__(self, config: SiglipTextConfig):
         self.embeddings = SiglipTextEmbeddings(config)
         self.encoder = SiglipEncoder(config)
         self.final_layer_norm = nn.LayerNorm(embed_dim, eps=config.layer_norm_eps)
-
-        self.head = nn.Linear(embed_dim, embed_dim)
+        if self.config.has_head:
+            self.head = nn.Linear(embed_dim, embed_dim)
```
Comment (NielsRogge) [inline src/transformers/models/siglip/modeling_siglip.py:930]:
```suggestion
        self.head = nn.Linear(embed_dim, embed_dim) if self.config.has_head else None
```

**Comment by merveenoyan** [comment]
@NielsRogge I confirmed image/text output embeds with atol of 1e-4 against JAX ones and all seems well, if you don't have any other comments we could merge 🤗

**Review by NielsRogge — APPROVED** [review]
LGTM, would be nice to upload the original checkpoints to a HF repo and leverage `hf_hub_download` so that anyone can run the conversion script.

**Comment by merveenoyan** [comment]
I uploaded tokenizers and currently uploading the models

File: src/transformers/models/siglip/modeling_siglip.py
Comment (ArthurZucker) [inline src/transformers/models/siglip/modeling_siglip.py:1]: We might want to add a new model with modular! it should be super super simple ! 🤗

File: src/transformers/models/siglip/convert_siglip_to_hf.py [hunk 15]
```diff
@@ -266,32 +277,49 @@ def convert_siglip_checkpoint(model_name, pytorch_dump_folder_path, verify_logit
...
+            if "img/Transformer/encoderblock/" in k:
+                original_array = np.array(v)
+                split_arrays = np.array_split(original_array, config.vision_config.num_hidden_layers, axis=0)  # 27
+                for i in range(config.vision_config.num_hidden_layers):
+                    new_key = re.sub(r"(encoderblock)/", rf"\1_{i}/", k)
+                    new_state_dict[f"params/{new_key}"] = split_arrays[i].squeeze()
+            else:
+                new_state_dict[f"params/{k}"] = v
+
+    state_dict = new_state_dict
```
Comment (ArthurZucker) [inline src/transformers/models/siglip/convert_siglip_to_hf.py:302]: let's add this to the new conversion script!

**Review by ArthurZucker — COMMENTED** [review]
Let's use `modular` !!! 🚀

File: src/transformers/models/siglip/modeling_siglip.py
Comment (merveenoyan) [reply src/transformers/models/siglip/modeling_siglip.py:1]: as much as I'm a huge huge fan of modular transformers, for this case I'd like to avoid because we do not need to have a separate class, this is a SigLIP model, and it's only one layer at the bottom which is 1-2 lines of code changing. I feel like it would overcomplicate in this case. WDYT?

File: src/transformers/models/siglip/convert_siglip_to_hf.py (hunk 15 above)
Comment (merveenoyan) [reply src/transformers/models/siglip/convert_siglip_to_hf.py:302]: @ArthurZucker architecture is not different here, it's only the way they put the encoder weights together which needed a modification before renaming the keys.

File: src/transformers/models/siglip/modeling_siglip.py
Comment (NielsRogge) [reply src/transformers/models/siglip/modeling_siglip.py:1]: I agree with @merveenoyan here since no new paper was introduced, it's still the same model/paper/code base

**Review by ArthurZucker — COMMENTED** [review]
Can you just initialize the weight of the Linear to be `eye` and no bias? This way it's more aligne with our philosophy!

File: src/transformers/models/siglip/configuration_siglip.py [hunk 16]
```diff
@@ -96,6 +98,7 @@ def __init__(
         pad_token_id=1,
         bos_token_id=49406,
         eos_token_id=49407,
+        has_head=True,
```
Comment (ArthurZucker) [inline src/transformers/models/siglip/configuration_siglip.py:101]: same here

File: src/transformers/models/siglip/modeling_siglip.py [hunk 17]
```diff
@@ -926,8 +926,11 @@ def __init__(self, config: SiglipTextConfig):
         self.embeddings = SiglipTextEmbeddings(config)
         self.encoder = SiglipEncoder(config)
         self.final_layer_norm = nn.LayerNorm(embed_dim, eps=config.layer_norm_eps)
-
-        self.head = nn.Linear(embed_dim, embed_dim)
+        self.head = (
+            nn.Linear(embed_dim, embed_dim, bias=False)
+            if config.has_head
+            else lambda x: x @ torch.eye(embed_dim, device=x.device)
+        )
```
Comment (ArthurZucker) [inline src/transformers/models/siglip/modeling_siglip.py:933]: No 😅 the idea is that you set the self.head's weights to `eye` in the conversion script. This way it's a no-op!

**Review by ArthurZucker — COMMENTED** [review]
🤗 sorry for the late review, the idea is to process the checkpoint!

**Comment by merveenoyan** [comment]
the latest ckpts I download from google cloud (in their official script) happen to have the same setup as the last one (no params/ prefix and different setup in encoder), I asked google folks to confirm. I'll have to change the script again if so. I have addressed Arthur's prev comments but haven't pushed yet since I want to confirm it works for older checkpoints.

**Comment by ArthurZucker** [comment]
Sounds good

**Comment by merveenoyan** [comment]
@NielsRogge I have pushed the final version with the changes Arthur asked, both models I converted match atol.

apparently there are some changes to keys and for some reason old models are not compatible with the code anymore (basically I downloaded and uploaded them using big_vision code). 
I am a bit too busy to debug and since we already have old models in place this doesn't seem urgent, can you check what's up?

**Review by ArthurZucker — COMMENTED** [review]
Looks good, thanks for updating, I think we can hold off a little bit to merge until this no longer breaks for old checkpoints! 🤗
